local count = require("@self/count")
local concat = require("@self/concat")
local concat_deep = require("@self/concat_deep")
local copy = require("@self/copy")
local copy_deep = require("@self/copy_deep")
local freeze_deep = require("@self/freeze_deep")
local filter = require("@self/filter")
local map = require("@self/map")
local pop = require("@self/pop")
local some = require("@self/some")
local partition = require("@self/partition")
local pick = require("@self/pick")
local omit = require("@self/omit")
local set = require("@self/set")
local insert = require("@self/insert")
local remove_index = require("@self/remove_index")
local move = require("@self/move")
local swap = require("@self/swap")

return table.freeze({
	count = count,
	concat = concat,
	concat_deep = concat_deep,
	copy = copy,
	copy_deep = copy_deep,
	freeze_deep = freeze_deep,
	filter = filter,
	map = map,
	pop = pop,
	some = some,
	partition = partition,
	pick = pick,
	omit = omit,
	set = set,
	insert = insert,
	remove_index = remove_index,
	move = move,
	swap = swap,
})
